### PR TITLE
Masked AES 128 bit, encrypt only (boolean mask technique) 

### PR DESCRIPTION
**Pull request by shelib01**

### CODE REVIEW

File: include/mbedtls/config.h [hunk 1]
```diff
@@ -654,6 +654,25 @@
  */
 //#define MBEDTLS_AES_SCA_COUNTERMEASURES
 
+/**
+ * \def MBEDTLS_AES_128_BIT_MASKED
+ *
+ * Requires MBEDTLS_AES_SCA_COUNTERMEASURES
```
Comment (AndrzejKurek) [inline include/mbedtls/config.h:660]: If this is required - please add a relevant check to check-config.h

File: include/mbedtls/config.h [hunk 2]
```diff
@@ -654,6 +654,25 @@
...
+ * Add boolean masking against possible combined side-channel-attack fault injection attacks.
+ *
+ * Uncommenting this macro adds data, key and Sbox masking additionally to dummy rounds
+ *
+ * Tradeoff:
+ * Uncommenting this macro does not increase codesize in MBEDTLS_AES_ROM_TABLES configuration.
+ * Uncommenting this macro increases codesize in AES RAM tables configuration by ~600 byte.
+ * The performance loss is ~50% with 128 bit AES encrypt.
+ *
+ * This option is dependent of \c MBEDTLS_ENTROPY_HARDWARE_ALT.
```
Comment (AndrzejKurek) [inline include/mbedtls/config.h:671]: Depends on? Or is independent of? If it depends on something, please add this information to the "required" line above.

File: include/mbedtls/config.h [hunk 3]
```diff
@@ -654,6 +654,25 @@
...
 //#define MBEDTLS_AES_SCA_COUNTERMEASURES
 
+/**
+ * \def MBEDTLS_AES_128_BIT_MASKED
+ *
+ * Requires MBEDTLS_AES_SCA_COUNTERMEASURES
+ *
+ * Add boolean masking against possible combined side-channel-attack fault injection attacks.
+ *
+ * Uncommenting this macro adds data, key and Sbox masking additionally to dummy rounds
```
Comment (AndrzejKurek) [inline include/mbedtls/config.h:664]:
```suggestion
 * Uncommenting this macro adds data, key and Sbox masking additionally to dummy rounds.
```

File: include/mbedtls/config.h [hunk 4]
```diff
@@ -654,6 +654,25 @@
...
+ *
+ * Tradeoff:
+ * Uncommenting this macro does not increase codesize in MBEDTLS_AES_ROM_TABLES configuration.
+ * Uncommenting this macro increases codesize in AES RAM tables configuration by ~600 byte.
```
Comment (AndrzejKurek) [inline include/mbedtls/config.h:668]:
```suggestion
 * Uncommenting this macro increases codesize in AES RAM tables configuration by ~600 bytes.
```

File: library/aes.c [hunk 5]
```diff
@@ -690,6 +721,8 @@ static void mbedtls_generate_fake_key( unsigned int keybits, mbedtls_aes_context
 /*
  * AES key schedule (encryption)
  */
+
```
Comment (AndrzejKurek) [inline library/aes.c:724]: Unnecessary whitespace.

File: library/aes.c [hunk 6]
```diff
@@ -1026,6 +1058,262 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+  //Calculate m1',m2',m3',m4'
+  calcMixColmask(mask);
+  flow_control++;
+
+  //Calculate the masked Sbox
+  if (calcSboxMasked(mask, sbox_masked) == 0){
+      flow_control++;
+  }
+
+#define MASK_INIT_CONTROL 19
```
Comment (AndrzejKurek) [inline library/aes.c:1219]: This define in the middle of a function is a bit strange.

File: library/aes.c [hunk 7]
```diff
@@ -1142,22 +1456,35 @@ int mbedtls_internal_aes_encrypt( mbedtls_aes_context *ctx,
...
     do
     {
         // Get pointer to the real or fake data
         aes_data_ptr = aes_data_table[round_ctrl_table[tindex] >> 4];
         offset = round_ctrl_table[tindex] & 0x04;
         stop_mark = round_ctrl_table[tindex] & 0x03;
-
+#if defined (MBEDTLS_AES_128_BIT_MASKED)
+        if (aes_masked_round( &aes_data_ptr->xy_values[0], aes_data_ptr->rk_ptr,
+                               aes_data_ptr->round, mask, sbox_masked) == 0)
```
Comment (AndrzejKurek) [inline library/aes.c:1484]: Please add braces for this complex if.

File: library/aes.c [hunk 8]
```diff
@@ -1167,15 +1494,24 @@ int mbedtls_internal_aes_encrypt( mbedtls_aes_context *ctx,
...
     } while( stop_mark == 0 );
 
     // Calculate final AES round + dummy rounds
     do
     {
         aes_data_ptr = aes_data_table[round_ctrl_table[tindex] >> 4];
         stop_mark = round_ctrl_table[tindex] & 0x03;
+#if defined (MBEDTLS_AES_128_BIT_MASKED)
+        if ( aes_masked_round_final( &aes_data_ptr->xy_values[0],
+                                     aes_data_ptr->rk_ptr, sbox_masked ) == 0)
```
Comment (AndrzejKurek) [inline library/aes.c:1510]: Please add braces for this complex if

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
I wasn't able to review the substantive part of this PR yet, but please fix the formatting:
- 4 spaces indentation;
- spaces after "(" and before ")", spaces after ",", spaces before and after "*" and other operations;
- maximum 80 characters per line (see for example line 1174); 
- no double empty lines;

File: library/aes.c [hunk 9]
```diff
@@ -1026,6 +1058,262 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+  (*state)[3][3] = (*state)[2][3];
+  (*state)[2][3] = (*state)[1][3];
+  (*state)[1][3] = temp;
+
+}
+
+#define mul_02(num) ( (num << 1) ^ (0x11b & -(num >> 7)) )
+#define mul_03(num) ( mul_02(num) ^ num )
+
+static void calcMixColmask(uint32_t mask[10])
```
Comment (AndrzejKurek) [inline library/aes.c:1150]: If this is for "calculating and mixing masked columns", then: 
```suggestion
static void calcMixColMask(uint32_t mask[10])
```

File: library/aes.c [hunk 10]
```diff
@@ -1026,6 +1058,262 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+
+
+static int init_masking_encrypt(const uint8_t *rk, uint8_t *rk_masked, uint32_t mask[10], uint8_t sbox_masked[256] )
+{
+  volatile int flow_control = 0;
+  unsigned int i = 0;
+
+  mbedtls_platform_memcpy(rk_masked, rk, MBEDTLS_AES_128_EXPANDED_KEY_SIZE_IN_WORDS*4);
+
+
```
Comment (AndrzejKurek) [inline library/aes.c:1202]: Double endline

File: library/aes.c [hunk 11]
```diff
@@ -1026,6 +1058,262 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+  }
+
+  if (i == 4) {
+        return 0;
+  }
+
+  return MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;
+}
+
+
```
Comment (AndrzejKurek) [inline library/aes.c:1194]: Double endline

Comment (shelib01) [reply library/aes.c:1198]: fixed

File: library/aes.c (hunk 10 above)
Comment (shelib01) [reply library/aes.c:1214]: Fixed

File: library/aes.c (hunk 9 above)
Comment (shelib01) [reply library/aes.c]: Fixed

**Comment by shelib01** [comment]
> I wasn't able to review the substantive part of this PR yet, but please fix the formatting:
> 
> * 4 spaces indentation;
> * spaces after "(" and before ")", spaces after ",", spaces before and after "*" and other operations;
> * maximum 80 characters per line (see for example line 1174);
> * no double empty lines;
> * use `return( value )`.

Fixed

File: include/mbedtls/config.h (hunk 1 above)
Comment (shelib01) [reply include/mbedtls/config.h]: Fixed

File: include/mbedtls/config.h (hunk 2 above)
Comment (shelib01) [reply include/mbedtls/config.h]: Fixed

File: include/mbedtls/config.h (hunk 3 above)
Comment (shelib01) [reply include/mbedtls/config.h]: Fixed

File: include/mbedtls/config.h (hunk 4 above)
Comment (shelib01) [reply include/mbedtls/config.h]: Fixed

File: library/aes.c (hunk 5 above)
Comment (shelib01) [reply library/aes.c]: Fixed

File: library/aes.c (hunk 6 above)
Comment (shelib01) [reply library/aes.c]: Fixed

File: library/aes.c (hunk 7 above)
Comment (shelib01) [reply library/aes.c]: Fixed

File: library/aes.c (hunk 8 above)
Comment (shelib01) [reply library/aes.c]: Fixed

File: library/aes.c [hunk 12]
```diff
@@ -1026,6 +1056,299 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+    (*state)[0][3] = (*state)[3][3];
+    (*state)[3][3] = (*state)[2][3];
+    (*state)[2][3] = (*state)[1][3];
+    (*state)[1][3] = temp;
+}
+
+#define mul_02( num ) ( ( num << 1 ) ^ ( 0x11b & - ( num >> 7 ) ) )
+#define mul_03( num ) ( mul_02( num ) ^ num )
+
+static void calc_mix_colmn_mask( uint32_t mask[10] )
```
Comment (AndrzejKurek) [inline library/aes.c:1150]: Is it intended that mask[4] and mask[5] aren't used here?

File: library/aes.c [hunk 13]
```diff
@@ -1026,6 +1056,299 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
...
+}
+
+//Calculate the the invSbox to change from Mask m to Mask m'
+static int calc_sbox_masked( uint32_t mask[10], uint8_t sbox_masked[256] )
+{
+    volatile unsigned int i = 0;
+
+    for ( i = 0; i < 256; i++ )
+    {
+        sbox_masked[i ^ mask[4]] = FSb[i] ^ mask[5];
```
Comment (AndrzejKurek) [inline library/aes.c:1165]: Are you sure that values above 255 can't be present in the mask variable?

**Review by AndrzejKurek — APPROVED** [review]
I have tested this PR with valgrind, and there are no leaks. 
There are some formatting errors still present, but nothing major.
I wasn't able to review the algorithm of masking present in this PR, but given the time constraints, I am accepting this PR.

File: library/aes.c (hunk 13 above)
Comment (shelib01) [reply library/aes.c:1165]: yes, I random it in range 1-255

**Pull request merged**